### ISSUE TITLE
Retriever Inference Seems to be Hanging on retriever.generate_question_vector

### ISSUE THREAD

**Issue by mosicr** [facebookresearch/DPR#124]
Hi, 
I generated dense embeddings via:
python3 generate_dense_embeddings.py model_file=/home/ranko_mosic/DPR/dpr/downloads/checkpoint/retriever/single/n\
   q/bert-base-encoder.cp ctx_src=dpr_wiki shard_id=0 num_shards=50 out_file=/home/ranko_mosic/DPR/retriever_inf

I am now trying to execute retriever inference via: 
python3 dense_retriever.py model_file=/home/ranko_mosic/DPR/dpr/downloads/checkpoint/retriever/single/nq/bert-base-\
 encoder.cp qa_dataset=nq_test ctx_datatsets=[dpr_wiki] encoded_ctx_files=[\"/home/ranko_mosic/DPR/test_retr/retriev\
 er_inf*\"] out_file=retr_inf.out

It seems to be hanging at dense_retriever.py , line 354
    questions_tensor = retriever.generate_question_vectors(
questions, query_token=qa_src.special_query_token
    )
It seems like query_token=qa_src.special_query_token is None
[2021-04-06 11:13:09,924][root][INFO] - Using special token None
Please advise.

**Comment by vlad-karpukhin** [comment]
"special_query_token is None" - this is a normal expected case, the issue must be due to some other reason.
Can you check your gpu utilization during generate_question_vectors?

**Comment by mosicr** [comment]
Thanks Vlad, makes sense as I am trying to execute on CPU actually :-)
Will try on GPU next. 
Much appreciated !!

**Comment by mosicr** [comment]
Hi, 
Ran on GPU ( Colab ), fails with:
NNNNNNNNNNNNNN Before special token
[2021-04-07 13:06:17,060][root][INFO] - Using special token None
Traceback (most recent call last):
  File "dense_retriever.py", line 360, in main
    questions, query_token=qa_src.special_query_token
  File "dense_retriever.py", line 129, in generate_question_vectors
    selector=self.selector,
  File "dense_retriever.py", line 95, in generate_question_vectors
    _, out, _ = question_encoder(q_ids_batch, q_seg_batch, q_attn_mask)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/content/DPR/dpr/models/hf_models.py", line 232, in forward
    attention_mask=attention_mask,
  File "/usr/local/lib/python3.7/dist-packages/transformers/modeling_bert.py", line 825, in forward
    input_ids=input_ids, position_ids=position_ids, token_type_ids=token_type_ids, inputs_embeds=inputs_embeds
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/transformers/modeling_bert.py", line 207, in forward
    inputs_embeds = self.word_embeddings(input_ids)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/sparse.py", line 158, in forward
    self.norm_type, self.scale_grad_by_freq, self.sparse)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py", line 1916, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Input, output and indices must be on the current device

**Comment by vlad-karpukhin** [comment]
DPR works only on GPU, we didn't even try CPUs, even though it has nocuda param, it actually makes no difference.

The error above suggests that some of the model& arguments are on different gpus.
Model Input data are explicitly moved to gpus memory here: https://github.com/facebookresearch/DPR/blob/master/dense_retriever.py#L78-L80

And then DP model wrapper (there is no need to use pytorch DDP launcher for dense_retriever) distributes it across all participating gpus.
Looks like there may be some issue with the model placement. Can you provide the full log so I can have a look at how it was initialized.
Overall, this is weird, I'm running dense_retriever al least several times per day on 2-gpu server without any issues like this.

**Comment by mosicr** [comment]
Thanks, Vlad, here is the full output ( unless you want me to set HYDRA_FULL_ERROR=1 ):
2021-04-07 13:11:34.843124: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2021-04-07 13:11:37,479 [INFO] faiss.loader: Loading faiss with AVX2 support.
2021-04-07 13:11:37,479 [INFO] faiss.loader: Loading faiss.
[2021-04-07 13:11:38,120][root][INFO] - args.local_rank -1
[2021-04-07 13:11:38,121][root][INFO] - WORLD_SIZE None
[2021-04-07 13:11:38,148][root][INFO] - Initialized host 18a50656820e as d.rank -1 on device=cuda, n_gpu=1, world size=1
[2021-04-07 13:11:38,148][root][INFO] - 16-bits training: False 
[2021-04-07 13:11:38,148][root][INFO] - CFG (after gpu  configuration):
[2021-04-07 13:11:38,155][root][INFO] - indexers:
  flat:
    _target_: dpr.indexer.faiss_indexers.DenseFlatIndexer
  hnsw:
    _target_: dpr.indexer.faiss_indexers.DenseHNSWFlatIndexer
  hnsw_sq:
    _target_: dpr.indexer.faiss_indexers.DenseHNSWSQIndexer
qa_dataset: nq_test
ctx_datatsets:
- dpr_wiki
encoded_ctx_files:
- /content/DPR/retriever_inf*
out_file: /content/DPR/retr_inf.out
match: string
n_docs: 100
validation_workers: 16
batch_size: 128
do_lower_case: true
encoder_path: null
index_path: null
kilt_out_file: null
model_file: /content/DPR/dpr/downloads/checkpoint/retriever/single/nq/bert-base-encoder.cp
validate_as_tables: false
rpc_retriever_cfg_file: null
indexer: flat
special_tokens: null
local_rank: -1
global_loss_buf_sz: 150000
device: cuda
distributed_world_size: 1
no_cuda: false
n_gpu: 1
fp16: false
fp16_opt_level: O1
encoder:
  encoder_model_type: hf_bert
  pretrained_model_cfg: bert-base-uncased
  pretrained_file: null
  projection_dim: 0
  sequence_length: 256
  dropout: 0.1
  fix_ctx_encoder: false
  pretrained: true
datasets:
  nq_test:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.nq-test
  nq_train:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.nq-train
  nq_dev:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.nq-dev
  trivia_test:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.trivia-test
  trivia_train:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.trivia-train
  trivia_dev:
    _target_: dpr.data.retriever_data.CsvQASrc
    file: data.retriever.qas.trivia-dev
ctx_sources:
  dpr_wiki:
    _target_: dpr.data.retriever_data.CsvCtxSrc
    file: data.wikipedia_split.psgs_w100
    id_prefix: 'wiki:'

[2021-04-07 13:11:38,156][root][INFO] - Reading saved model from /content/DPR/dpr/downloads/checkpoint/retriever/single/nq/bert-base-encoder.cp
[2021-04-07 13:11:38,736][root][INFO] - model_state_dict keys odict_keys(['model_dict', 'optimizer_dict', 'scheduler_dict', 'offset', 'epoch', 'encoder_params'])
[2021-04-07 13:11:50,881][root][INFO] - Selecting standard question encoder
*****************
cfg.device cuda
[2021-04-07 13:11:52,839][root][INFO] - Loading saved model state ...
[2021-04-07 13:11:52,839][root][INFO] - Encoder state prefix question_model.
[2021-04-07 13:11:52,935][root][INFO] - Encoder vector_size=768
[2021-04-07 13:11:52,935][root][INFO] - qa_dataset: nq_test
[2021-04-07 13:11:52,959][dpr.data.download_data][INFO] - Requested resource from https://dl.fbaipublicfiles.com/dpr/data/retriever/nq-test.qa.csv
[2021-04-07 13:11:52,959][dpr.data.download_data][INFO] - Download root_dir /content/DPR
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - File to be downloaded as /content/DPR/downloads/data/retriever/qas/nq-test.csv
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - File already exist /content/DPR/downloads/data/retriever/qas/nq-test.csv
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - Loading from https://dl.fbaipublicfiles.com/dpr/nq_license/LICENSE
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - File already exist /content/DPR/downloads/data/retriever/qas/LICENSE
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - Loading from https://dl.fbaipublicfiles.com/dpr/nq_license/README
[2021-04-07 13:11:52,960][dpr.data.download_data][INFO] - File already exist /content/DPR/downloads/data/retriever/qas/README
[2021-04-07 13:11:53,013][root][INFO] - Index class <class 'dpr.indexer.faiss_indexers.DenseFlatIndexer'> 
NNNNNNNNNNNNNN Before special token
[2021-04-07 13:11:53,014][root][INFO] - Using special token None
Traceback (most recent call last):
  File "dense_retriever.py", line 358, in main
    questions, query_token=qa_src.special_query_token
  File "dense_retriever.py", line 129, in generate_question_vectors
    selector=self.selector,
  File "dense_retriever.py", line 95, in generate_question_vectors
    _, out, _ = question_encoder(q_ids_batch, q_seg_batch, q_attn_mask)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/content/DPR/dpr/models/hf_models.py", line 232, in forward
    attention_mask=attention_mask,
  File "/usr/local/lib/python3.7/dist-packages/transformers/modeling_bert.py", line 825, in forward
    input_ids=input_ids, position_ids=position_ids, token_type_ids=token_type_ids, inputs_embeds=inputs_embeds
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/transformers/modeling_bert.py", line 207, in forward
    inputs_embeds = self.word_embeddings(input_ids)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/sparse.py", line 158, in forward
    self.norm_type, self.scale_grad_by_freq, self.sparse)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py", line 1916, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Input, output and indices must be on the current device

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

**Comment by vlad-karpukhin** [comment]
Hi @mosicr, do you still experience this issue?

**Comment by mosicr** [comment]
Hi Vlad, 
I didn't have a chance to retest.

**Comment by nikoletta-toth** [comment]
Hi,

I have the same problem with this error msg.

1. I train your model from scratch with my own dataset, and the training + eval at every epoch is going fine until the 30. epoch, then at the 31. epoch fails with the above error. This is super weird, because if everything is fine until 30 epoch, then why does it fail at 31 suddenly?

2. I am also trying to fine-tune your checkpoint with my dataset, and I get the same error.
After loading in the **checkpoint.retriever.single-adv-hn.nq.bert-base-encoder,** the training starts with the 31. epoch, and at that point when the validation comes, the whole thing fails with that error msg.

3. Setup for the fine-tuning:
AWS p3.2xlarge instance with one 16 GB GPU,
I clone your repository from git, then
set up the configs: my dataset, the epoch number, and the model_file with your checkpoint, the device to be cuda
Then I start the training and it just fails.

4. So the problem I believe is that somewhere something is stuck on the CPU and other stuff is on the GPU. I also sometimes get the error msg: 
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and CPU!
I am trying to figure out where and what to send either to GPU or CPU so that everything is on the same device, but no luck yet.

Do you have an idea to solve this?

Thanks!

**Comment by vlad-karpukhin** [comment]
Hi,

I don't see your message on the github for some reason.
I replied to your another issue's comment that you probably should try to
disable optimizer & lr scheduler loading when downloading from the provided
checkpoint.
set ignore_checkpoint_optimizer=true as the command line argument.
Ideally, you'd want to load the optimizer but ignore the lr scheduler for
the model finetuning, but I don't have time to integrate this better
approach to the public repo now. (the logic is here:
https://github.com/facebookresearch/DPR/blob/1ee31c6c5335f05c45138161eb5ce2c03023115d/train_dense_encoder.py#L594
).
Also, in case you have your custom datasets, it is recommended to train the
model from raw bert using a combination of datasets (all provided +
your own) adjusting each dataset sampling rates. I found this scheme
working slightly but consistently better than the finetuning.

Best regards,
Vladimir
